src/utils/url.js

```diff
@@ -67,11 +90,40 @@ const mimes = {
 export const getMimeTypeFromFileExtension = (extension: string): string =>
   mimes[extension.toLowerCase()] || 'application/octet-stream';
 
-export const autocompleteUrl = (value: string = '', protocol: string, append: string): string =>
-  value.length > 0
-    ? `${hasProtocol(value) ? '' : protocol}${value || 'your-org'}${
-        value.indexOf('.') === -1 ? append : ''
-      }`
-    : '';
-
 export const isValidUrl = (url: string): boolean => urlRegex({ exact: true }).test(url);
+
+export type AutocompletionData = {|
+  protocol: string,
+  hostname: string,
+|};
+
+export type AutocompletionPieces = [string | null, string, string | null];
+
+/**
+ * Given user input purporting to be a Zulip server's host name, provide
+ * a prefix, derived value, and suffix which may suffice to turn it into
+ * a real URL.
+ *
+ * Presently, the derived value will always be equal to the input value;
+ * this should not be relied on, as it may change in future.
+ */
+export const autocompleteHostnamePieces = (
```

Comment:
I don't think "hostname" is the right term to use here. The input might
* be an IP address, with no hostname present
* contain a port
* contain a scheme (and `//`)

But what these (and the hostname case) all have in common is that they're fragments of a URL.

So I think "autocomplete URL" is the right central words to describe what this is doing -- it's taking a fragment of a URL (that contains the most salient bits), and filling it out into a real, complete URL.

Of course it's not just any URL, and the UI context that makes this particular autocompletion helpful wouldn't make sense for lots of kinds of URLs. It's really about specifically autocompleting a Zulip organization URL, or "realm" URL as it's often called in the API and code.

So a good more specific name might be `autocompleteRealmUrl` (and `...Pieces`).

Reply:
The autocompletion mechanism is theoretically somewhat more general than just Zulip realms, but it *is* very strongly realm-oriented. Arguably, I should have hardcoded `zulipchat.com` at this level – I left it general because it started that way, but I don't think this particular bit of configurability will ever buy anyone anything.

At any rate, I'll use that name (or `autocompleteRealm`, possibly).